### PR TITLE
Car logging

### PR DESCRIPTION
**Pull request by shrimalmadhur**
- Logging logic is now removed from RecordingFile. Instead I pass the PawnWrapper which gets the logging line from the respective pawn. This will help up extend the logging and make it easy for different Pawn

### CODE REVIEW

File: Unreal/Plugins/AirSim/Source/Car/CarPawn.cpp [hunk 1]
```diff
@@ -182,28 +182,28 @@ ACarPawn::ACarPawn()
...
     Vehicle4W->MinNormalizedTireLoad = 0.0f;
-    Vehicle4W->MinNormalizedTireLoadFiltered = 0.2308f;
+    Vehicle4W->MinNormalizedTireLoadFiltered = 0.2f;
     Vehicle4W->MaxNormalizedTireLoad = 2.0f;
     Vehicle4W->MaxNormalizedTireLoadFiltered = 2.0f;
 
     // Engine 
     // Torque setup
-    Vehicle4W->EngineSetup.MaxRPM = 5700.0f;
+    Vehicle4W->MaxEngineRPM = 5700.0f;
```
Comment (mitchellspryn) [inline Unreal/Plugins/AirSim/Source/Car/CarPawn.cpp]: Why did you change this? I think this will cause the max engine RPM to be set incorrectly for default pawn

File: Unreal/Plugins/AirSim/Source/Car/CarPawn.cpp [hunk 2]
```diff
@@ -182,28 +182,28 @@ ACarPawn::ACarPawn()
 
     // Adjust the tire loading
     Vehicle4W->MinNormalizedTireLoad = 0.0f;
-    Vehicle4W->MinNormalizedTireLoadFiltered = 0.2308f;
+    Vehicle4W->MinNormalizedTireLoadFiltered = 0.2f;
```
Comment (mitchellspryn) [inline Unreal/Plugins/AirSim/Source/Car/CarPawn.cpp]: Why?

File: Unreal/Plugins/AirSim/Source/Car/CarPawn.cpp [hunk 3]
```diff
@@ -305,7 +304,7 @@ void ACarPawn::initializeForBeginPlay(bool enable_rpc, const std::string& api_se
     InternalCamera4 = this->GetWorld()->SpawnActor<APIPCamera>(pip_camera_class_, camera_transform, camera_spawn_params);
     InternalCamera4->AttachToComponent(InternalCameraBase4, FAttachmentTransformRules::KeepRelativeTransform);
     InternalCamera5 = this->GetWorld()->SpawnActor<APIPCamera>(pip_camera_class_, FTransform(FRotator(0, 180, 0), FVector::ZeroVector), camera_spawn_params);
-    InternalCamera5->AttachToComponent(InternalCameraBase5, FAttachmentTransformRules::KeepRelativeTransform);
+    InternalCamera5->AttachToComponent(InternalCameraBase4, FAttachmentTransformRules::KeepRelativeTransform);
```
Comment (mitchellspryn) [inline Unreal/Plugins/AirSim/Source/Car/CarPawn.cpp]: Why did you change this? I think this will cause the backup camera to be in the wrong place

File: Unreal/Plugins/AirSim/Source/Car/CarPawn.cpp [hunk 4]
```diff
@@ -416,10 +415,10 @@ void ACarPawn::setupInputBindings()
     UAirBlueprintLib::BindAxisToKey(FInputAxisKeyMapping("MoveRight", EKeys::Gamepad_LeftX, 1), this,
         this, &ACarPawn::MoveRight);
 
-    UAirBlueprintLib::BindAxisToKey(FInputAxisKeyMapping("MoveForward", EKeys::Gamepad_RightTriggerAxis, 1), this,
+    UAirBlueprintLib::BindAxisToKey(FInputAxisKeyMapping("MoveForward", EKeys::Gamepad_RightY, -1), this,
```
Comment (mitchellspryn) [inline Unreal/Plugins/AirSim/Source/Car/CarPawn.cpp]: Why change all these key bindings?

File: Unreal/Plugins/AirSim/Source/Car/CarPawn.cpp [hunk 5]
```diff
@@ -615,4 +619,29 @@ void ACarPawn::UpdatePhysicsMaterial()
...
+std::string ACarPawn::getLogString()
+{
+    // Timestamp \t Speed \t Throttle \t Steering \t Brake \t gear
+
+    // timestamp
+    uint64_t timestamp_millis = static_cast<uint64_t>(msr::airlib::ClockFactory::get()->nowNanos() / 1.0E6);
+
+    // Speed
+    float KPH = FMath::Abs(GetVehicleMovement()->GetForwardSpeed()) * 0.036f;
+    int32 KPH_int = FMath::FloorToInt(KPH);
```
Comment (mitchellspryn) [inline Unreal/Plugins/AirSim/Source/Car/CarPawn.cpp]: What is the 0.036f? I guess a conversion factor, perhaps a const float is in order?

File: Unreal/Plugins/AirSim/Source/Car/SimModeCar.cpp [hunk 6]
```diff
@@ -29,7 +29,8 @@ void ASimModeCar::BeginPlay()
 
     createVehicles(vehicles_);
 
-    columns = { "Timestamp", "Position(x)", "Position(y)" , "Position(z)", "Orientation(w)", "Orientation(x)", "Orientation(y)", "Orientation(z)", "ImageName" };
+    // Timestamp \t Speed \t Throttle \t Steering \t Brake \t gear
```
Comment (mitchellspryn) [inline Unreal/Plugins/AirSim/Source/Car/SimModeCar.cpp]: This comment is confusing and contradicts the next line (it doesn't have ImageName). I'd remove it.

File: Unreal/Plugins/AirSim/Source/Multirotor/SimModeWorldMultiRotor.cpp [hunk 7]
```diff
@@ -148,6 +146,36 @@ void ASimModeWorldMultiRotor::setupVehiclesAndCamera(std::vector<VehiclePtr>& ve
...
+        .append(std::to_string(kinematics->pose.orientation.z())).append("\t");
+
+    return line;
+
+    //std::stringstream ss;
+    //ss << timestamp_millis << "\t";
+    //ss << kinematics.pose.position.x() << "\t" << kinematics.pose.position.y() << "\t" << kinematics.pose.position.z() << "\t";
+    //ss << kinematics.pose.orientation.w() << "\t" << kinematics.pose.orientation.x() << "\t" << kinematics.pose.orientation.y() << "\t" << kinematics.pose.orientation.z() << "\t";
+    //ss << "\n";
+    //return ss.str();
```
Comment (mitchellspryn) [inline Unreal/Plugins/AirSim/Source/Multirotor/SimModeWorldMultiRotor.cpp]: Do we still need this block commented code?

File: Unreal/Plugins/AirSim/Source/Recording/RecordingFile.cpp [hunk 8]
```diff
@@ -11,7 +11,7 @@ RecordingFile::RecordingFile(std::vector <std::string> columns)
 {
     this->columns = columns;
 }
-void RecordingFile::appendRecord(TArray<uint8>& image_data, const msr::airlib::Kinematics::State* kinematics)
+void RecordingFile::appendRecord(TArray<uint8>& image_data, VehiclePawnWrapper* wrapper)
```
Comment (mitchellspryn) [inline Unreal/Plugins/AirSim/Source/Recording/RecordingFile.cpp]: can VehiclePawnWrapper be const?

File: Unreal/Plugins/AirSim/Source/VehiclePawnWrapper.h [hunk 9]
```diff
@@ -65,6 +65,10 @@ class VehiclePawnWrapper
     const GeoPoint& getHomePoint() const;
     const CollisionInfo& getCollisonInfo() const;
 
+    std::string logLine;
```
Comment (mitchellspryn) [inline Unreal/Plugins/AirSim/Source/VehiclePawnWrapper.h]: Should this be private? Otherwise, why have getters / setters?

Also, with naming convention, should it be "log_line_"?

**Review by mitchellspryn — CHANGES_REQUESTED** [review]
Quick review

File: Unreal/Plugins/AirSim/Source/Car/CarPawn.cpp (hunk 5 above)
Comment (shrimalmadhur) [reply Unreal/Plugins/AirSim/Source/Car/CarPawn.cpp]: 0.036f is the conversion from cm/s to km/h

File: Unreal/Plugins/AirSim/Source/Car/SimModeCar.cpp (hunk 6 above)
Comment (shrimalmadhur) [reply Unreal/Plugins/AirSim/Source/Car/SimModeCar.cpp]: Good catch. will make this change.

File: Unreal/Plugins/AirSim/Source/Multirotor/SimModeWorldMultiRotor.cpp (hunk 7 above)
Comment (shrimalmadhur) [reply Unreal/Plugins/AirSim/Source/Multirotor/SimModeWorldMultiRotor.cpp]: This was there before. @sytelus do we need this commented block?

File: Unreal/Plugins/AirSim/Source/VehiclePawnWrapper.h (hunk 9 above)
Comment (shrimalmadhur) [reply Unreal/Plugins/AirSim/Source/VehiclePawnWrapper.h]: makes sense. Will make it private

File: Unreal/Plugins/AirSim/Source/Car/CarPawn.cpp (hunk 2 above)
Comment (shrimalmadhur) [reply Unreal/Plugins/AirSim/Source/Car/CarPawn.cpp]: I think this change happened unintentionally during merge. will refer to master and make sure to make it consistent

File: Unreal/Plugins/AirSim/Source/Car/CarPawn.cpp (hunk 1 above)
Comment (shrimalmadhur) [reply Unreal/Plugins/AirSim/Source/Car/CarPawn.cpp]: I think this change happened unintentionally during merge. will refer to master and make sure to make it consistent

File: Unreal/Plugins/AirSim/Source/Car/CarPawn.cpp (hunk 3 above)
Comment (shrimalmadhur) [reply Unreal/Plugins/AirSim/Source/Car/CarPawn.cpp]: I think this change happened unintentionally during merge. will refer to master and make sure to make it consistent

File: Unreal/Plugins/AirSim/Source/Car/CarPawn.cpp (hunk 4 above)
Comment (shrimalmadhur) [reply Unreal/Plugins/AirSim/Source/Car/CarPawn.cpp]: I think this change happened unintentionally during merge. will refer to master and make sure to make it consistent

File: Unreal/Plugins/AirSim/Source/Recording/RecordingFile.cpp (hunk 8 above)
Comment (shrimalmadhur) [reply Unreal/Plugins/AirSim/Source/Recording/RecordingFile.cpp]: I don't think so. the member functions get/set logline changes the variable logLine which can't be made const
